### PR TITLE
Ui protocol spike

### PR DESCRIPTION
**Pull request by hkgumbs**

### CODE REVIEW

File: Sources/Core/Player.swift [hunk 1]
```diff
@@ -1,6 +1,6 @@
 protocol Player {
-    var maker: String { get }
-    init(marker: String)
+    //let maker: String { get }
+    //init(marker: String)
     func getMove(avaliableMoves: [Int]) -> Int
     func getMarker() -> String
```
Comment (hkgumbs) [inline Sources/Core/Player.swift]: why did you change this from an attribute to a method? I think the attribute is more informative to clients, since it suggests that you want the result to be non-computed.

File: Sources/Specs/main.swift [hunk 2]
```diff
@@ -2,3 +2,4 @@ import XTest
 
 Suite(groups: BoardSpec()).run()
 Suite(groups: EndGameDetectorSpec()).run()
+Suite(groups: CommandLineUISpec()).run()
```
Comment (hkgumbs) [inline Sources/Specs/main.swift]: https://github.com/Connor-Vickers/SwiftTackToe/commit/e3a0aad0bb94595915f2feb66cddfdcb744ea41e#commitcomment-17860593

File: Sources/Core/Human.swift [hunk 3]
```diff
@@ -1,7 +1,26 @@
...
+    func getMove(avaliableMoves: [Int]) -> Int{
+        var move = 0
+        func isValidMove(input:String?) -> Bool{
+            if (input == nil){
+                return false
+            }
+            if (Int(input!) == nil){
+                return false
+            }
+            move = Int(input!)!
```
Comment (hkgumbs) [inline Sources/Core/Human.swift]: I realize that you are checking for nils here, but you should almost never forcibly unwrap optionals in swift. Look up `guard` statements and "nil coalescing."

File: Sources/Core/CommandLineUI.swift [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    public init(){}
+    public func send(message: String){
+        print(message)
+    }
+    public func get()->String{
+        return readLine(strippingNewline: true)!
+    }
+}
+
+public class TestIO: IO{
```
Comment (hkgumbs) [inline Sources/Core/CommandLineUI.swift]: Nice!!!!
Nitpick, but I prefer to keep test-doubles with my tests (i.e. in the `Specs` directory)

File: Sources/Core/CommandLineUI.swift [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        }
+    }
+}
+
+public protocol IO {
+    func send(message: String)
+    func get()->String
+}
+
+public class CommandLine: IO{
```
Comment (hkgumbs) [inline Sources/Core/CommandLineUI.swift]: This doesn't need to be a class. Always prefer value types.

File: Sources/Core/Human.swift [hunk 6]
```diff
@@ -1,7 +1,26 @@
...
+        self.marker = marker
+        self.ui = ui
     }
-    func getMove(avaliableMoves: [Int]) -> Int
-    func getMarker() -> String
-}*/
\ No newline at end of file
+    func getMove(avaliableMoves: [Int]) -> Int{
+        var move = 0
+        func isValidMove(input:String?) -> Bool{
```
Comment (Jgoodrich07) [inline Sources/Core/Human.swift]: This is not idiomatic Swift. It would be better to do something like `let validate: (String?) -> Bool =  {input in ...}` or pass it in by name as a static function.
